src/ray/gcs/gcs_client/gcs_client.cc

```diff
@@ -191,6 +186,9 @@ Status PythonGcsClient::Connect(const ClusterID &cluster_id,
         return HandleGcsError(reply.status());
       }
       std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+      channel_ =
+          rpc::GcsRpcClient::CreateGcsChannel(options_.gcs_address_, options_.gcs_port_);
+      node_info_stub_ = rpc::NodeInfoGcsService::NewStub(channel_);
```

Comment:
What's the root cause?

Reply:
The channel was not being recreated during retries, so an RPC that failed the first time was doomed to fail for every subsequent retry (spaced 1s apart) since a faulty channel would be returned. We also don't check whether the channel is faulty when we first create it.

Reply:
removed

Reply:
Can we just use https://github.com/ray-project/ray/pull/39017 (cc @rynewang)?

Reply:
My understanding is that the linked PR is for connection fluctuations on a working channel, whereas in this case, it's not clear that the channel created is working at all (so it is necessary to create a new one). i.e. RPCs fail on the bad channel for 5 seconds, whereas the server comes up within a second and a half -- so wait_for_ready will either be slower than recreating, or it will never work.

Reply:
I see, that makes sense. So this is the case retry is better than waiting until channel is recovered from a bad state. 

Bad channel -> is it because the channel may never recover because of GCS restart?

Reply:
Not sure whether the state on the channel is laggy (by several seconds), or if the created channel is just permanently bad (maybe because GCS was not available at time of creation, although in this case it's still starting up, not restarting)

Reply:
So the laggy case could be handled by wait_for_ready (the transient state should be recovered), but the closed channel (without proper shutdown) cannot be handled by this IIUC? Anyway it makes sense to retry here.

Btw, @rynewang don't we still need wait_for_ready here still at least for a few seconds in case it takes time to establish the channel? It seems like this approach won't work if the channel connection cannot be estabilshed within 1 second?

Reply:
This confuses me a bit. My understanding is a channel can always be created, albeit in a TRANSIENT_FAILURE state if remote is not accessible, and will retry to connect if in that state according to the [grpc doc](https://grpc.github.io/grpc/cpp/md_doc_connectivity-semantics-and-api.html). What does it mean by `since a faulty channel would be returned` and `doomed to fail for every subsequent retry`? I guess with [a wait for retry PR](https://github.com/ray-project/ray/pull/39017) it should work.

Reply:
This makes sense to me. Btw, what's the reason why the first channel is not ready in the first place? Is it because we try connecting to GCS before it is ready? (and exponential backoff will make it longer time to make transient failure -> ready state?)